### PR TITLE
Add `MinkowskiMetric` abstract type

### PR DESCRIPTION
**Pull request by eliascarv**
This new type will allow third-party packages to differentiate Minkowski metrics at the type level. It will also allow third-party packages to implement their own Minkowski metrics.

A practical example of a package that could use this new abstract type is NearestNeighbors.jl, which is currently using a union type for this purpose:

https://github.com/KristofferC/NearestNeighbors.jl/blob/master/src/kd_tree.jl#L1

With this new type NearestNeighbors.j could use Minkowski metrics defined in any package.

### CODE REVIEW

**Comment by juliohm** [comment]
Appreciate if a maintainer can review and merge this.

**Comment by juliohm** [comment]
ping @KristofferC @dkarrasch

File: src/metrics.jl
Comment (dkarrasch) [inline src/metrics.jl:16]: What do we need this for? Can't we replace any usage of `UnionMinkowskiMetric` by `MinkowskiMetric`? Users could simply subtype `MinkowskiMetric`, which is already abstract.

**Review by dkarrasch — COMMENTED** [review]
This doesn't seem to hurt anyone, so this LGTM, except for the consideration of adding two types vertically into the type hierarchy.

File: src/metrics.jl
Comment (eliascarv) [reply src/metrics.jl:16]: I created this type to follow the pattern of the other "Union" types:
```
abstract type UnionPreMetric <: PreMetric end

abstract type UnionSemiMetric <: SemiMetric end

abstract type UnionMetric <: Metric end
```

From what I understand, these "Union" types are treated differently in the implementation.

However, since all subtypes of `MinkowskiMetric` were subtypes of `UnionMetric`, it is possible to replace `UnionMinkowskiMetric` with `MinkowskiMetric` in the code.

Comment (dkarrasch) [reply src/metrics.jl:16]: It's been a very long while when I was looking at/dealing with these things, but these types are needed for multiple dispatch AFAIR. They used to be just `Union`s, but then people couldn't reuse their infrastructure. Basically, the corresponding *metrics have the same computational pattern, just different operations in the details.

Couldn't you just subtype `MinkowskiMetric <: UnionMetric`? Thinking about it more, I think you actually should do that, because otherwise these types fall out of the `UnionMetric` type branch, which potentially other users rely on. You would then simply introduce another type layer, to distinguish some metrics (`MinkowskiMetric`) from other `UnionMetric` metrics.

Comment (eliascarv) [reply src/metrics.jl:16]: So, is the PR code correct? Should we keep `UnionMinkowskiMetric`?

Comment (eliascarv) [reply src/metrics.jl]: Done!

**Review by dkarrasch — COMMENTED** [review]
Looks much cleaner now and non-breaking.

**Review by dkarrasch — APPROVED** [review]
LGTM

**Comment by eliascarv** [comment]
@nalimilan, this PR is ready. Can you review it please?

File: src/metrics.jl
Comment (nalimilan) [inline src/metrics.jl:15]: Make this a docstring (with appropriate formatting)? It would be useful to explain how this differs from `Minkowski` as it could confuse users. Could also be interesting to say what values of `p` subtypes correspond to.

Also:
```suggestion
# a Minkowski metric is a metric that is defined by the formula:
```

Comment (eliascarv) [reply src/metrics.jl]: Done!

Comment (eliascarv) [reply src/metrics.jl]: @nalimilan, do you think this new type should be exported?

Comment (nalimilan) [reply src/metrics.jl]: Good question. In theory I would say yes, but the fact that `MinkowskiMetric <: UnionMetric` is problematic to me, as the latter is internal and its interface isn't documented (see https://github.com/JuliaStats/Distances.jl/issues/151#issuecomment-612460855 for how it apparently works).

Actually this raises a more fundamental point: in the current state of the PR, if people want to implement a `MinkowskiMetric` subtype, they have to implement the `UnionMetric` interface, or at least define enough methods so that none of the fallback methods defined for that type are used. This isn't logical. It would seem better to introduce `UnionMinkowskiMetric <: MinkowskiMetric <: PreMetric` and redefine `UnionMetrics` to include that (see https://github.com/JuliaStats/Distances.jl/pull/163 for context about `UnionMetric`). What do you think @dkarrasch?

Comment (eliascarv) [reply src/metrics.jl]: @dkarrasch, the PR was like this at the beginning:

```julia
# /src/generic.jl
abstract type MinkowskiMetric <: Metric end

# /src/metrics.jl
abstract type UnionMinkowskiMetric <: MinkowskiMetric end
```

But after some discussions this was changed to the current code.

Comment (nalimilan) [reply src/metrics.jl]: Ah sorry I had missed that discussion. This is definitely tricky. @dkarrasch You said:

> Couldn't you just subtype `MinkowskiMetric <: UnionMetric`? Thinking about it more, I think you actually should do that, because otherwise these types fall out of the `UnionMetric` type branch, which potentially other users rely on.

But why do you think it would be a problem that these types would not be `UnionMetric`? This is internal anyway. It seems to me that according to your reasoning, we could drop all `Union*` types, e.g. replacing `UnionMetric` with just `Metric`.

Comment (dkarrasch) [reply src/metrics.jl]: After second thought, I think @nalimilan's proposal (which seems like the original state of the PR) is better. I confused `UnionMetric` with `UnionMetrics`, and what I was having in mind as the `UnionMetric` mechanism/infrastructure (the `eval_*` stuff) is actually a  `UnionMetrics` mechanism. Sorry for spreading confusion, and thanks @nalimilan for your care- and thoughtful review!

Comment (dkarrasch) [reply src/metrics.jl]: @eliascarv Can you bump the patch version right in this PR? Then after we merge, we can release right away.

Comment (eliascarv) [reply src/metrics.jl]: @nalimilan, so we can export the type `MinkowskiMetric` (not `UnionMinkowskiMetric`), right?

Comment (nalimilan) [reply src/metrics.jl]: Yes I think so.

File: src/generic.jl
Comment (nalimilan) [inline src/generic.jl:25]:
```suggestion
    MinkowskiMetric <: Metric
```

File: src/Distances.jl
Comment (nalimilan) [inline src/Distances.jl:32]: Intended change?

Comment (eliascarv) [reply src/Distances.jl:32]: No, it must have been the autoformat, undoing

**Review by dkarrasch — APPROVED** [review]

**Pull request merged**
